pkg/secureenclave/secureenclave.m

```diff
@@ -0,0 +1,271 @@
+  }
+
+  if (keyAttrs){
+    CFRelease((CFTypeRef)keyAttrs);
+    keyAttrs = NULL;
+  }
+
+  return res;
+}
+
+size_t createKey(unsigned char** ret, char** retErr){
+    CFErrorRef error = NULL;
+
+    // ignoring the depreciated warning for "kSecAttrAccessibleAlwaysThisDeviceOnly" becuase this is the setting
+    // we need to be able to access secure enclave for this particualar key even when device is locked
+    // this was depreciated without providing an alternative
+    // it's expected that a new key would need to be generated on a full re-install of launcher that removes the launcher.db
+    #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    SecAccessControlRef access = SecAccessControlCreateWithFlags(kCFAllocatorDefault,
+                                                                 kSecAttrAccessibleAlwaysThisDeviceOnly,
+                                                                 kSecAccessControlPrivateKeyUsage,
+                                                                 &error);
+
+    if (error) {
+        CFStringRef errStr = CFErrorCopyDescription(error);
+        CFRelease(error);
+        error = NULL;
+        *retErr = CFStringToCString(errStr);
+        CFRelease(errStr);
+        errStr = NULL;
+        return 0;
+    }
+
+    NSDictionary* attributes =
+    @{ (id)kSecAttrKeyType: (id)kSecAttrKeyTypeECSECPrimeRandom,
+        (id)kSecAttrKeySizeInBits: @256,
+        (id)kSecAttrTokenID: (id)kSecAttrTokenIDSecureEnclave,
+        (id)kSecPrivateKeyAttrs:
+        @{  (id)kSecAttrIsPermanent: @YES,
+            (id)kSecAttrAccessControl: (id)access,
+        },
+    };
+
+    // create the private key
+    SecKeyRef privateKey = SecKeyCreateRandomKey((__bridge CFDictionaryRef)attributes, &error);
+
+    CFRelease(access);
+    access = NULL;
+    CFRelease(attributes);
+    attributes = NULL;
+
+    if ((error) || (!privateKey)) {
+        CFStringRef errStr = NULL;
+        if (!error){
+          errStr = CFSTR("no private key or error returned by SecKeyCreateRandomKey");
+        } else {
+          errStr = CFErrorCopyDescription(error);
+          CFRelease(error);
+          error = NULL;
+        }
+
+        *retErr = CFStringToCString(errStr);
+        CFRelease(errStr);
+        errStr = NULL;
+        return 0;
+    }
+
+    // get public key from private key
+    SecKeyRef publicKey = SecKeyCopyPublicKey(privateKey);
+    CFRelease(privateKey);
+    privateKey = NULL;
+
+    if (!publicKey) {
+      CFStringRef errStr = CFSTR("public key was not able to be derived from private key by SecKeyCopyPublicKey");
+      *retErr = CFStringToCString(errStr);
+      CFRelease(errStr);
+      errStr = NULL;
+      return 0;
+    }
+
+    // extract just the public key data
+    CFDataRef publicKeyRef = ExtractPubKey(publicKey);
+    *ret = CFDataToUint8(publicKeyRef);
+    CFIndex size = CFDataGetLength(publicKeyRef);
+    CFRelease(publicKeyRef);
```

Comment:
This could be NULL since `ExtractPubKey` can return NULL

Reply:
now null checking